Benchmark changes

inflector opened the pull request.

inflector commented on opencog/benchmark/python_diary.txt:
```diff
@@ -422,3 +422,78 @@ Add nodes - ConceptNode sugar - Scheme        111.437µs           8,973
+
+This is 25 times slower than it was. Yikes! Need to dig into the reason why.
+
+
```
@linas Any idea why this slowed down?

inflector commented on opencog/benchmark/python_diary.txt:
```diff
@@ -422,3 +422,78 @@ Add nodes - ConceptNode sugar - Scheme        111.437µs           8,973
+Get Outgoing - no temporary list                1.627µs         614,720
+Bind - stub_bindlink - Cython                   0.376µs       2,662,691
+Bind - bindlink - Cython                      360.240µs           2,775
+Test scheme_eval_h(+ 2 2)                      76.558µs          13,062
```
@linas this is the overhead measure for calling scheme from Python that is slow. This is not a dig on the speed of scheme, just the current implementation of the way scheme_eval_h works.

linas commented:
re bindlink performance, please realize its a two-step process: a "compile" step, and a "run" step.  The compile step parses the bindlink, and is slow/heavyweight -- it creates various C++ structs and fills them in.  The second, "run" step should be fast. The bindlink, once compiled, can be used repeatedly; it doesn't need to be recompiled with each use.

Its possible that somehow you are now measuring the compile step, whereas before you were not?

inflector commented on opencog/benchmark/python_diary.txt:
```diff
@@ -422,3 +422,78 @@ Add nodes - ConceptNode sugar - Scheme        111.437µs           8,973
+Bind - bindlink - Cython                      360.240µs           2,775
+Test scheme_eval_h(+ 2 2)                      76.558µs          13,062
+Bind - cog-bind - Scheme                      474.588µs           2,107
+Add nodes - cog-new-node - Scheme             127.043µs           7,871
```
@linas Compare this to the 7.5 microseconds for adding directly through the Python bindings.

inflector commented on opencog/benchmark/python_diary.txt:
```diff
@@ -422,3 +422,78 @@ Add nodes - ConceptNode sugar - Scheme        111.437µs           8,973
+Add nodes - ConceptNode sugar - Scheme        110.855µs           9,020
+Predicates - get_predicates                     2.260µs         442,523
+Predicates - get_predicates_for                 1.909µs         523,728
+Predicates - cog-get-pred - Scheme            130.991µs           7,634
```
@linas Here is another comparing the Python get_predicates against the same functionality called through scheme_eval_h. This is a measure of the transition using scheme_eval_h, not the speed of the scheme code versus Python, they're both wrapping the C++ thinly.

Once I fix scheme_eval_h to not lose state, i.e. using your idea of reentrant scheme, this won't be so different.

linas commented:
re: bindlink performance -- there is nothing I can think of, in the last year, that should have changed bindlink performance in a negative way.  Whatever the change is, its surprising.

inflector commented:
I didn't actually change the test. It's still just calling into the C++ bindlink routine, the same one that cog-bind calls.

Where can I learn more about the bind and compile process?

The specific test is this code:

```
def prep_bind(atomspace):
    for scheme_file in scheme_preload:
        load_scm(atomspace, scheme_file)
    scheme_eval(atomspace, "(use-modules (opencog))")
    scheme_eval(atomspace, "(use-modules (opencog query))")

    # Define several animals and something of a different type as well
    scheme_animals = \
        '''
        (InheritanceLink (ConceptNode "Frog") (ConceptNode "animal"))
        (InheritanceLink (ConceptNode "Zebra") (ConceptNode "animal"))
        (InheritanceLink (ConceptNode "Deer") (ConceptNode "animal"))
        (InheritanceLink (ConceptNode "Spaceship") (ConceptNode "machine"))
        '''
    scheme_eval_h(atomspace, scheme_animals)

def prep_bind_python(atomspace):
    prep_bind(atomspace)

    # Define a graph search query
    bind_link_query = \
        '''
        (BindLink
            ;; The variable to be bound
            (VariableNode "$var")
            ;; The pattern to be searched for
            (InheritanceLink
                (VariableNode "$var")
                (ConceptNode "animal")
            )
            ;; The value to be returned.
            (VariableNode "$var")
        )
        '''
    return scheme_eval_h(atomspace, bind_link_query)
```

followed by the test which is a loop calling `bindlink` with `prep_atom` being the atom returned by evaluating the `bind_link_query`:
```
def test_bind(atomspace, prep_atom):
    n = 10000
    for i in xrange(n):
        result = bindlink(atomspace, prep_atom)
    return n
```

linas commented:
Hm. OK. The compile step happens when the bindlink is first added to the atomspace.  It appears that `prep_bind_python` does this.  Once you have the handle to it, the compile step is over.  The `test_bind` function is just running the query, it should be (should have been) fast.

linas commented:
I'm vaguely fearful that the protoatom work of #513 causes this, but if so, it would be a major surprise.

inflector commented:
I'll take a look at the protoatom changes tomorrow and see if anything stands out.

linas commented:
Bisecting would be an easier way of finding the problem.  I find it hard to believe that protoatom would do this; it was measured carefully and repeatedly, and makes only a "minor" impact on the usual benchmarks.  So I find it hard to believe that it is the source of the difference.

bisect first, then ponder what bisection reveals.

inflector commented:
​Well, I wanted to take a look at protoatom anyway, so no time wasted. Will
begin bisecting and will let you know.​

On Sat, Feb 6, 2016 at 5:03 PM, Linas Vepštas <notifications@github.com>
wrote:

> Bisecting would be an easier way of finding the problem. I find it hard to
> believe that protoatom would do this; it was measured carefully and
> repeatedly, and makes only a "minor" impact on the usual benchmarks. So I
> find it hard to believe that it is the source of the difference.
>
> bisect first, then ponder what bisection reveals.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/atomspace/pull/628#issuecomment-180872832>.
>

inflector commented:
@linas 

Here's what I found. Pull request #1522 was the one that caused the slowdown.

```
    								   git commit
Date                                         time in µs         ops/sec
----								   ----------------------------------------
Sep 9, 2015                            63b584da7abfdc89095ddd89f05661f63fb8b4a8
Bind - bindlink - Cython                      350.940µs           2,849

Wed Jul 15 08:24:37 2015 +0200         2c687a41db1e05144a6509371f4a13fc6fa36f3e
Bind - bindlink - Cython                      351.737µs           2,843

Tue May 5 17:50:04 2015 -0500		   3e5451ab302d190c8435ce63ea8cac1b72af741f
Bind - bindlink - Cython                      318.244µs           3,142

Wed Apr 15 10:23:18 2015 -0700         38a7e06068d8fada1d03e1c16430e12aa4ebf89f
Bind - bindlink - Cython                      344.461µs           2,903

Tue Apr 14 16:51:46 2015 +0200         ff8646a9372daef68da69e843719f1b72fdc1659
Bind - bindlink - Cython                      350.990µs           2,849

Merge pull request #1522 from linas/equ
Sun Apr 12 21:26:57 2015 -0500         af094d399e342cc2542905dd4f2d030ab038f717
Bind - bindlink - Cython                      360.898µs           2,770
NOTE: removed #define DEBUG 1 in PatternMatchCallback.h

<<< SLOWS DOWN HERE

Merge pull request #1521 from linas/pln
Sat Apr 11 03:15:49 2015 -0500         126281142fdd6632cbb87cb26bade5ed83ae395a
Bind - bindlink - Cython                       27.713µs          36,084
NOTE: removed #define DEBUG 1 in PatternMatchCallback.h

Thu Apr 9 21:04:31 2015 -0500          3650ce5d00cba8e29f121283642e19c662d5e4c1
Bind - bindlink - Cython                       26.243µs          38,105
NOTE: had to #undef DEBUG in a few placed to get rid of spurious prints and
remove #define DEBUG 1 in PatternMatchCallback.h

Thu Mar 26 19:33:59 2015 -0500         2495d5df5b35d1e41bb6decf600b13efaa4eb42f
Bind - bindlink - Cython                       19.270µs          51,893
```

linas commented:
I'm confused.  You did use `git bisect` to find this, right? what did git bisect report?

linas commented:
I mean, the before and after commits are both merge commits .. I thought git-bisect doesn't report merges, that instead, it will give you the actual commit on the actual branch that the change is on...

linas commented:
If you did use git bisect, can you post `git bisect log` ?

inflector commented:
I didn't use bisect as most of the inter-merge commits don't build. Believe
me, the hard part is getting the build to work reliably and not getting
spurious prints or build bugs which were later fixed (often one or two
commits later). Bisect just does the math for you. That's not the hard
part. You still have to manually build and test it yourself with bisect.

Getting the old builds working is non-trivial at times. Only the merge
commits appeared to represent reasonably consistent points where cmake ..
and make might work with a decent chance of success.

A point of caution if you try to reproduce this. don't do this in a vm
where you have your code in:

/home/opencog/atomspace

the old make file does a regex on the opencog/atomspace and assumes that's
the internal atomspace directory and omits the class server references in
the type construction macro.

It's searching for "opencog/atomspace" and will match the above directory
against all the subdirectories instead of just the atomspace directory,
which in the above case would be /home/opencog/atomspace/opencog/atomspace

I fixed this by not having my outer-level atomspace directory use that name:

/home/opencog/old_atomspace/

- Curtis

On Tue, Feb 9, 2016 at 8:56 PM, Linas Vepštas <notifications@github.com>
wrote:

> If you did use git bisect, can you post git bisect log ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/atomspace/pull/628#issuecomment-182165792>.
>

inflector commented:
@linas

Digging into the individual commits, commit https://github.com/opencog/atomspace/commit/13be76e0f24603c43fe67a745da473fad97e570e caused the slowdown. Click on the link >> https://github.com/opencog/atomspace/commit/13be76e0f24603c43fe67a745da473fad97e570e << to see the diff.

The potential significant difference seems to be whether or not the temp atomspace is a stack variable or a class member variable. The diff itself is a fairly small set of changes.

Details on the bisect of commit timings between pull request #1521 and pull request #1522 below:
```
Merge pull request #1522 from linas/equ
Sun Apr 12 21:26:57 2015 -0500         af094d399e342cc2542905dd4f2d030ab038f717
Bind - bindlink - Cython                      360.898µs           2,770
NOTE: removed #define DEBUG 1 in PatternMatchCallback.h

More testing
Sun Apr 12 12:19:24 2015 -0500         f7da1ea83354cdac468b022a38402ee4b6232c76
Bind - bindlink - Cython                      351.237µs           2,847
NOTE: removed #define DEBUG 1 in PatternMatchCallback.h

clauses at top now work
Sun Apr 12 10:21:25 2015 -0500         e9268fa7852f5302d2345382c7249c1726be5370
Bind - bindlink - Cython                      353.114µs           2,831
NOTE: removed #define DEBUG 1 in PatternMatchCallback.h

SLOWS DOWN HERE >>>

Remove the old virtual-link API
Sun Apr 12 08:04:44 2015 -0500         13be76e0f24603c43fe67a745da473fad97e570e
Bind - bindlink - Cython                      345.920µs           2,890
NOTE: removed #define DEBUG 1 in PatternMatchCallback.h

<<< SLOWS DOWN HERE

Ongoing attempts to fix circular dependency
Sun Apr 12 07:33:22 2015 -0500         98e3d2c5c331d695765f800ab6037d0bb220d8a5
Bind - bindlink - Cython                       27.910µs          35,829
NOTE: removed #define DEBUG 1 in PatternMatchCallback.h

Add the evaluatable and executable maps
Sat Apr 11 11:56:03 2015 -0500		   c20a6ec1c587b21b34a23b3c86dae87c50f75828
Bind - bindlink - Cython                       26.651µs          37,522
NOTE: removed #define DEBUG 1 in PatternMatchCallback.h

Merge pull request #1521 from linas/pln
Sat Apr 11 03:15:49 2015 -0500         126281142fdd6632cbb87cb26bade5ed83ae395a
Bind - bindlink - Cython                       27.713µs          36,084
NOTE: removed #define DEBUG 1 in PatternMatchCallback.h
```

linas commented:
Thanks for narrowing it down.  But, as I look at the diff, it is highly counter-intuitive.  The goal of pulling it off the stack, and putting it into the object was to remove the overhead of running the constructors repeatedly, on the stack -- the constructors would run once, and that's it. That this would cause a 10x slowdown is remarkable.  I'm not sure I want to look at this today; I want to make forward progress on something else.   We ned to convert this conversation into an issue, so that its not forgotten.

linas commented:
I opened #634 to track this.
